api/paidAction/zap.js

```diff
@@ -0,0 +1,151 @@
+  } else if (actIds) {
+    acts = await tx.itemAct.findMany({ where: { id: { in: actIds } }, include: { item: true } })
+  } else {
+    throw new Error('No invoice or actIds')
+  }
+
+  const msats = acts.reduce((a, b) => a + BigInt(b.msats), BigInt(0))
+  const sats = msatsToSats(msats)
+  const itemAct = acts.find(act => act.act === 'TIP')
+
+  // give user and all forwards the sats
+  await tx.$executeRaw`
+    WITH forwardees AS (
+      SELECT "userId", ((${itemAct.msats}::BIGINT * pct) / 100)::BIGINT AS msats
+      FROM "ItemForward"
+      WHERE "itemId" = ${itemAct.itemId}::INTEGER
+    ), total_forwarded AS (
+      SELECT COALESCE(SUM(msats), 0) as msats
+      FROM forwardees
+    ), forward AS (
+      UPDATE users
+      SET msats = users.msats + forwardees.msats
+      FROM forwardees
+      WHERE users.id = forwardees."userId"
+    )
+    UPDATE users
+    SET msats = msats + ${itemAct.msats}::BIGINT - (SELECT msats FROM total_forwarded)::BIGINT
+    WHERE id = ${itemAct.item.userId}::INTEGER`
+
+  // perform denomormalized aggregates: weighted votes, upvotes, msats, lastZapAt
+  // NOTE: for the rows that might be updated by a concurrent zap, we use UPDATE for implicit locking
+  await tx.$executeRaw`
+    WITH zapper AS (
+      SELECT trust FROM users WHERE id = ${itemAct.userId}::INTEGER
+    ), zap AS (
+      INSERT INTO "ItemUserAgg" ("userId", "itemId", "zapSats")
+      VALUES (${itemAct.userId}::INTEGER, ${itemAct.itemId}::INTEGER, ${sats}::INTEGER)
+      ON CONFLICT ("itemId", "userId") DO UPDATE
+      SET "zapSats" = "ItemUserAgg"."zapSats" + ${sats}::INTEGER, updated_at = now()
+      RETURNING ("zapSats" = ${sats}::INTEGER)::INTEGER as first_vote,
+        LOG("zapSats" / GREATEST("zapSats" - ${sats}::INTEGER, 1)::FLOAT) AS log_sats
+    )
+    UPDATE "Item"
+    SET
+      "weightedVotes" = "weightedVotes" + (zapper.trust * zap.log_sats),
+      upvotes = upvotes + zap.first_vote,
+      msats = "Item".msats + ${msats}::BIGINT,
+      "lastZapAt" = now()
+    FROM zap, zapper
+    WHERE "Item".id = ${itemAct.itemId}::INTEGER`
+
+  // record potential bounty payment
+  // NOTE: we are at least guaranteed that we see the update "ItemUserAgg" from our tx so we can trust
+  // we won't miss a zap that aggregates into a bounty payment, regardless of the order of updates
+  await tx.$executeRaw`
+    WITH bounty AS (
+      SELECT root.id, "ItemUserAgg"."zapSats" >= root.bounty AS paid, "ItemUserAgg"."itemId" AS target
+      FROM "ItemUserAgg"
+      JOIN "Item" ON "Item".id = "ItemUserAgg"."itemId"
+      LEFT JOIN "Item" root ON root.id = "Item"."rootId"
+      WHERE "ItemUserAgg"."userId" = ${itemAct.userId}::INTEGER
+      AND "ItemUserAgg"."itemId" = ${itemAct.itemId}::INTEGER
+      AND root."userId" = ${itemAct.userId}::INTEGER
+      AND root.bounty IS NOT NULL
+    )
+    UPDATE "Item"
+    SET "bountyPaidTo" = array_remove(array_append(array_remove("bountyPaidTo", bounty.target), bounty.target), NULL)
```

Comment:
We should've done this triple array call before. It basically makes `bountyPaidTo` into a set.